src/controller_examples/rabbitmq_controller/resource/headless_service.rs

```diff
@@ -0,0 +1,96 @@
+
+impl ResourceBuilder for HeadlessServiceBuilder {
+    fn get_request(rabbitmq: &RabbitmqCluster) -> KubeGetRequest {
+        KubeGetRequest {
+            api_resource: Service::api_resource(),
+            name: make_headless_service_name(rabbitmq),
+            namespace: rabbitmq.namespace().unwrap(),
+        }
+    }
+
+    fn make(rabbitmq: &RabbitmqCluster, state: &RabbitmqReconcileState) -> Result<DynamicObject, RabbitmqError> {
+        Ok(make_headless_service(rabbitmq).marshal())
+    }
+
+    fn update(rabbitmq: &RabbitmqCluster, state: &RabbitmqReconcileState, obj: DynamicObject) -> Result<DynamicObject, RabbitmqError> {
+        let service = Service::unmarshal(obj);
+        if service.is_ok() {
+            Ok(update_headless_service(rabbitmq, service.unwrap()).marshal())
+        } else {
+            Err(RabbitmqError::Error)
+        }
+    }
+
+    fn state_after_create_or_update(obj: DynamicObject, state: RabbitmqReconcileState) -> (res: Result<RabbitmqReconcileState, RabbitmqError>) {
+        let service = Service::unmarshal(obj);
+        if service.is_ok() {
+            Ok(RabbitmqReconcileState {
+                reconcile_step: RabbitmqReconcileStep::AfterKRequestStep(ActionKind::Get, ResourceKind::Service),
+                ..state
+            })
+        } else {
+            Err(RabbitmqError::Error)
+        }
+    }
+
+    open spec fn get_request_view(rabbitmq: RabbitmqClusterView) -> GetRequest {
+        GetRequest { key: resource::make_headless_service_key(rabbitmq) }
+    }
+
+    open spec fn make_view(rabbitmq: RabbitmqClusterView, state: RabbitmqReconcileStateView) -> Result<DynamicObjectView, RabbitmqError> {
+        Ok(resource::make_headless_service(rabbitmq).marshal())
+    }
+
+    open spec fn update_view(rabbitmq: RabbitmqClusterView, state: RabbitmqReconcileStateView, obj: DynamicObjectView) -> Result<DynamicObjectView, RabbitmqError> {
+        let service = ServiceView::unmarshal(obj);
+        if service.is_Ok() {
+            Ok(resource::update_headless_service(rabbitmq, service.get_Ok_0()).marshal())
+        } else {
+            Err(RabbitmqError::Error)
+        }
+    }
+
+    open spec fn state_after_create_or_update_view(obj: DynamicObjectView, state: RabbitmqReconcileStateView) -> (res: Result<RabbitmqReconcileStateView, RabbitmqError>) {
+        let service = ServiceView::unmarshal(obj);
+        if service.is_Ok() {
+            Ok(RabbitmqReconcileStateView {
+                reconcile_step: RabbitmqReconcileStep::AfterKRequestStep(ActionKind::Get, ResourceKind::Service),
+                ..state
+            })
+        } else {
+            Err(RabbitmqError::Error)
+        }
+    }
+
+    open spec fn resource_state_matches(rabbitmq: RabbitmqClusterView, obj: DynamicObjectView) -> bool {
+        &&& ServiceView::unmarshal(obj).is_Ok()
+        &&& ServiceView::unmarshal(obj).get_Ok_0().spec == resource::make_headless_service(rabbitmq).spec
```

Comment:
We probably don't want to talk about the entire spec of a service, since the clusterIP(s) is not decided by the controller (tho we haven't specified the built-in controller for service yet).